lte/gateway/c/session_manager/SetMessageManagerHandler.cpp

```diff
@@ -72,50 +72,95 @@ void SetMessageManagerHandler::SetAmfSessionContext(
   // Print the message from AMF
   PrintGrpcMessage(static_cast<const google::protobuf::Message&>(request_cpy));
   response_callback(Status::OK, SmContextVoid());
-  /*The Event Based main_thread invocation and runs to handle session state*/
-  m5g_enforcer_->get_event_base().runInEventBaseThread(
-      [this, response_callback, request_cpy]() {
-        // extract values from proto
-        auto imsi = request_cpy.common_context().sid().id();
-        std::string dnn =
-            request_cpy.common_context().apn();  // may not required for demo-1
-        std::string session_id = id_gen_.gen_session_id(imsi);
-
-        MLOG(MDEBUG) << "Requested session from UE with IMSI: " << imsi
-                     << " Generated sessioncontext ID" << session_id;
-        /*reach complete message from proto message*/
-        SessionConfig cfg = m5g_build_session_config(request_cpy);
-
-        /*check if it's initial message*/
-        if ((cfg.rat_specific_context.m5gsm_session_context().rquest_type() ==
-             INITIAL_REQUEST) &&
-            (cfg.common_context.sm_session_state() == CREATING_0)) {
-          /* it's new UE establisment request and need to create the session
-           * context
-           */
-          MLOG(MDEBUG)
-              << "AMF request type INITIAL_REQUEST and session state CREATING";
-          /* Read the SessionMap from global session_store,
-           * if it is not found, it will be added w.r.t imsi
-           */
+

```

Comment:
I think `SetAmfSessionContext` is doing to many stuff in the main thread. We should receive the grpc message and immediately place it on a `event_base`. 

Right now it seems the grpc server may be blocked while you process the message and decide what to do with it. In your previous implementation all the logic was under ` m5g_enforcer_->get_event_base().runInEventBaseThread` which may help in case high number of request come at the same time

Reply:
Can u more specific what u intent to say ?  Are you suggesting to remove some of the MLOG messages here or something else ?  Please be specific ?

Reply:
Please, check the previous implementation. Note you had 
```
m5g_enforcer_->get_event_base().runInEventBaseThread
```
at the very top. That means we create an event an run the rest inside the event.  

On the new code there is some logic that happens outside the event. In order to make sure as much code is run inside on an event, you should include everything you can inside the `m5g_enforcer_->get_event_base().runInEventBaseThread`, including the logs


(code below is not proper , just a sample)
```
m5g_enforcer_->get_event_base().runInEventBaseThread(
      [this, response_callback, request_cpy]() {
  // extract values from proto
  auto imsi       = request_cpy.common_context().sid().id();
  std::string dnn = request_cpy.common_context().apn();
  if (cfg.common_context.sm_session_state() == RELEASED_4) {
    if (cfg.common_context.sm_session_version() != 0) {
      MLOG(MERROR) << "Wrong version received from AMF for IMSI " << imsi
                   << " but continuing release request";
    }
    MLOG(MINFO) << "Release request for session from IMSI: " << imsi << " DNN "
                << dnn;
          auto session_map = session_store_.read_sessions({imsi});
          initiate_release_session(session_map, dnn, imsi);
    }
  } else {
    // The Event Based main_thread invocation and runs to handle session state
      std::string session_id = id_gen_.gen_session_id(imsi);
      MLOG(MINFO) << "Requested session from UE with IMSI: " << imsi
                  << " Generated session " << session_id;

      /* Message may be intial or modification message. Only taken care
       * intial message. Check if it's initial message
       */
      if ((cfg.rat_specific_context.m5gsm_session_context().rquest_type() ==
           INITIAL_REQUEST) &&
          (cfg.common_context.sm_session_state() == CREATING_0)) {
        /* it's new UE establisment request and need to create the session
         * context
         */
        MLOG(MINFO)
            << "AMF request type INITIAL_REQUEST and session state CREATING";
        /* Read the SessionMap from global session_store,
         * if it is not found, it will be added w.r.t imsi
         */
        auto session_map = session_store_.read_sessions({imsi});
        send_create_session(session_map, imsi, session_id, cfg, dnn);
      }
	}
 response_callback(Status::OK, SmContextVoid());
});
....    

```

Reply:
Changed the code.